### PR TITLE
Add more overload-related metrics

### PR DESCRIPTION
**Pull request by psarna**
This miniseries adds metrics which can help the users detect potential overloads:
 * due to having too many in-flight hints
 * due to exceeding the capacity of the read admission queue, on replica side

### CODE REVIEW

File: database.cc [hunk 1]
```diff
@@ -481,6 +486,11 @@ database::setup_metrics() {
                                        " to be able to admit new ones, if there is a shortage of permits."),
                        {streaming_label_instance}),
 
+        sm::make_derive("reads_shed_due_to_overload", _read_concurrency_sem.get_stats().total_reads_shed_due_to_overload,
+                       sm::description("The number of reads shed because the admission queue reached its max capacity."
+                                       " When the queue is full, excessive reads are shed to avoid overload."
+                       {streaming_label_instance}),
+
```
Comment (avikivity) [inline database.cc:493]: Unrelated: looks like there is duplication, we can extract the function to create metrics, and call it twice.

File: database.cc [hunk 2]
```diff
@@ -481,6 +486,11 @@ database::setup_metrics() {
                                        " to be able to admit new ones, if there is a shortage of permits."),
                        {streaming_label_instance}),
 
+        sm::make_derive("reads_shed_due_to_overload", _read_concurrency_sem.get_stats().total_reads_shed_due_to_overload,
```
Comment (avikivity) [inline database.cc:489]: Should this be _streaming_concurrency_semaphore? based on the label below.

File: database.cc [hunk 3]
```diff
@@ -456,6 +456,11 @@ database::setup_metrics() {
                                        " to be able to admit new ones, if there is a shortage of permits."),
                        {user_label_instance}),
 
+        sm::make_derive("reads_shed_due_to_overload", _read_concurrency_sem.get_stats().total_reads_shed_due_to_overload,
+                       sm::description("The number of reads shed because the admission queue reached its max capacity."
+                                       " When the queue is full, excessive reads are shed to avoid overload."
+                       {user_label_instance}),
+
         sm::make_gauge("active_reads", [this] { return max_count_streaming_concurrent_reads - _streaming_concurrency_sem.available_resources().count; },
```
Comment (avikivity) [inline database.cc:464]: @amnonh we should drop the exception logged on reader concurrency semaphore overloaded, and replace it with an alert on this metric

File: database.cc (hunk 2 above)
Comment (psarna) [reply database.cc:489]: Yes it should... Similarly, the system metrics below is also wrong. I'll fix that tomorrow

Comment (psarna) [reply database.cc:489]: today*

**Comment by psarna** [comment]
I pushed v2 with fixed semaphore references

**Comment by avikivity** [comment]
Dequeued.

**Comment by psarna** [comment]
update: I repushed a proper version that actually compiles and works

**Pull request closed without merging**